cirq/ops/n_qubit_diagonal_gate.py

```diff
@@ -0,0 +1,188 @@
+        if args.precision is not None:
+            rounded_angles = rounded_angles.round(args.precision)
+        if len(rounded_angles) <= 4:
+            diag_str = 'diag({})'.format(', '.join(proper_repr(angle) for angle in rounded_angles))
+        else:
+            diag_str = ', '.join(proper_repr(angle) for angle in rounded_angles[:2])
+            diag_str += ', ..., '
+            diag_str += ', '.join(proper_repr(angle) for angle in rounded_angles[-2:])
+            diag_str = 'diag({})'.format(diag_str)
+        return protocols.CircuitDiagramInfo(
+            [diag_str] + ['#' + str(i) for i in range(2, self._num_qubits_() + 1)]
+        )
+
+    def __pow__(self, exponent: Any) -> 'DiagonalGate':
+        if not isinstance(exponent, (int, float, sympy.Basic)):
+            return NotImplemented
+        angles = []
+        for angle in self._diag_angles_radians:
+            mulAngle = protocols.mul(angle, exponent, NotImplemented)
+            angles.append(mulAngle)
+        return DiagonalGate(angles)
+
+    def _value_equality_values_(self) -> Any:
+        return tuple(self._diag_angles_radians)
+
+    def _decompose_for_basis(self, index, bit_flip, theta, qubits):
+        if index == 0:
+            return []
+        largest_digit = self._num_qubits_() - (len(bin(index)) - 2)
+        yield common_gates.rz(2 * theta)(qubits[largest_digit])
+        _flip_bit = self._num_qubits_() - bit_flip - 1
+        if _flip_bit < largest_digit:
+            yield common_gates.CNOT(qubits[largest_digit], qubits[_flip_bit])
+        elif _flip_bit > largest_digit:
+            yield common_gates.CNOT(qubits[_flip_bit], qubits[largest_digit])
+
+    def _decompose_(self, qubits):
+        """Decompose the n-qubit diagonal gates into CNOT and Rz gates.
+
+        A 3 qubits decomposition looks like
+        0: ───────────────────────────────────X───Rz(6)───X───Rz(7)───X───Rz(5)───X───Rz(4)───
+                                              │           │           │           │
+        1: ───────────X───Rz(3)───X───Rz(2)───@───────────┼───────────@───────────┼───────────
+                      │           │                       │                       │
+        2: ───Rz(1)───@───────────@───────────────────────@───────────────────────@───────────
+
+        where the angles in Rz gates are corresponding to the fast-walsh-Hadamard transfrom
+        of diagonal_angles in the Gray Code order.
+
+        For n qubits decomposition looks similar but with 2^n-1 Rz gates and 2^n-2 CNOT gates.
+
+        The algorithm is implemented according to the paper:
+            Welch, Jonathan, et al. "Efficient quantum circuits for diagonal unitaries without
+            ancillas." New Journal of Physics 16.3 (2014): 033040.
+            https://iopscience.iop.org/article/10.1088/1367-2630/16/3/033040/meta
+        """
+        n = self._num_qubits_()
+        hat_angles = _fast_walsh_hadamard_transform(self._diag_angles_radians) / (2 ** n)
```

Comment:
I wonder how this behaves when the gate is parametrized. I think it would blow up. To avoid this we should add here: 

```
        if protocols.is_parameterized(self):
            return NotImplemented
```

Also, we'll need tests for that.

Reply:
Yeah, the current implementation cannot work with the parameterized cases. (Actually, it is still possible as long as we make symbol-based Hardmard transform.) I added the NotImplemented case. One thing I felt strange is

```
cirq.decompose(diagonal_gate(*cirq.LineQubit.range(2))) == []  # <-- parameterized gate
```

Note the result is [] instead of raising the error. Is that expected?

Reply:
That is not my expectation - I'd expect `[DiagonalGate(qubits)]` - i.e. the undecomposed op ... Maybe it's a bug in the decompose protocol?

Reply:
The issue is that you're `return`ing in a generator. When `decompose` encounters this, it attempts to iterate through the generator, encounters the return, and stops without `yield`ing anything.

See [this SO question](https://stackoverflow.com/questions/26595895/return-and-yield-in-the-same-function) for more information. TL;DR: the value of a `return` is ignored in a generator.

Reply:
Great catch, thank you! 🙇‍♂️

Reply:
Then, what should be the expected behavior we want here?

Some options:
1. Raise the error directly?
2. Make everything return a list instead of yield
3. Support the parameterized decomposition. It is easy to support except for one piece: GlobalPhaseOperation. It does not support sympy.exp(1j*xx) as input due to the unity check. [Awkward]

Reply:
Option 2 is the right solution!